src/test/christophermanahan/github/io/carnitas/EchoServerTest.java

```diff
@@ -0,0 +1,101 @@
+package christophermanahan.github.io.carnitas;
+
+import org.junit.jupiter.api.Test;
+
+import java.io.ByteArrayInputStream;
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.net.ServerSocket;
+import java.net.Socket;
+import java.util.Iterator;
+import java.util.List;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+class EchoServerTest {
+
+  @Test
+  void dataReceivedFromInputIsSentToOutputUntilDisconnect() throws IOException {
+    List<String> incomingData = List.of("data", "data");
+    TestSocket socket = new TestSocket(incomingData);
+    TestServerSocket serverSocket = new TestServerSocket(socket);
+    EchoServer echoServer = new EchoServer(serverSocket);
+
+    echoServer.run();
+
+    String outgoingData = socket.getOutputStream().toString();
+    List<String> outgoingDataList = List.of(outgoingData.split("\n"));
+    assertEquals(incomingData, outgoingDataList);
+  }
+
+  @Test void socketIsClosedWhenStreamIsDisconnected() throws IOException {
+    List<String> incomingData = List.of();
+    TestSocket socket = new TestSocket(incomingData);
+    TestServerSocket serverSocket = new TestServerSocket(socket);
+    EchoServer echoServer = new EchoServer(serverSocket);
+
+    echoServer.run();
+
+    assertTrue(socket.closed);
+  }
```

Comment:
Both of these tests have a similar setup. We could move some of this to some `@Before` method.